genomeGraph/genomeGraph.go

```diff
@@ -48,45 +49,61 @@ type Edge struct {
 func Read(filename string) *GenomeGraph {
 	simpleReader := fileio.NewByteReader(filename)
 	genome := EmptyGraph()
-	var currNode, homeNode, destNode *Node
 	var weight float32
 	var line string
 	var words []string = make([]string, 0, 2)
-	var seqIdx, homeNodeIdx, destNodeIdx uint32
+	var nodeId, homeNodeIdx, destNodeIdx uint32
+	var homeNode, destNode *Node
```

Comment:
This relates to one of my answers above.  In graph-like data structures (e.g. trees, our genome graph), they are typically assembled by having a unit (e.g. Node) that occurs many times to make up the larger data structure (e.g. graph).  These small units point to each other so that you may travel between them (e.g. edges in the graph).  It is important that each Node occur only once in memory, and then is accessed with pointers.  There is not really a way around using pointers (or inventing something that acts as a pointer) for these types of data structures.
We now have a sneaky trick in the graph where we define the nodes once without a pointer (in the Nodes slice), but this can only happen once in the whole program, or you would end up with multiple copies of node 17.  Multiple copies would both take more memory and mean that updating that node would only change one of the many copies of it, which would get confusing.  I would be happy to draw this out in person on a whiteboard, or give it a shot over google slides.

Reply:
Thanks Craig. That makes sense now that you explained it. We talked during a graph meeting once about having all unique nodes at their own addresses (ie a memory address (say 201) for a Node with Node.seq = CG and then any time a node of CG is needed we would point to that 201 memory address) and the graph being written with pointers. But I think at this moment it is written where the foundational graph has a memory address for each node (regardless if there sequence is unique to all nodes of the graph, so each CG node would have unique memory addresses) and then any time we are referencing a node from the main one we are using pointers. Correct me if I have something wrong.

Reply:
I think there is a subtle distinction to make here between a node and a sequence.  Not all GC nodes could be the same because the node also contains edges and if we made every GC node the same, they would all have to point to the same "next" and "prev" nodes, which would not work.  We could store the sequence "GC" only once, and have every GC node point to the same sequence of "GC" in memory.